File: lib/models/api_source.dart
import '../l10n/app_localizations.dart';

enum ApiSource {
  googleBooks,
  openLibrary,
  wikipedia,
  jikeFree,
}

class ApiSourceInfo {
  final ApiSource id;
  final String displayName;
  final String description;
  final bool enabledByDefault;
  final bool requiresKey;
  final String baseUrl;

  const ApiSourceInfo({
    required this.id,
    required this.displayName,
    required this.description,
    required this.enabledByDefault,
    required this.requiresKey,
    required this.baseUrl,
  });

  String localizedName(AppLocalizations loc) {
    switch (id) {
      case ApiSource.googleBooks:
        return loc.settings_google_title;
      case ApiSource.openLibrary:
        return loc.settings_openlibrary_title;
      case ApiSource.wikipedia:
        return loc.settings_wikipedia_title;
      case ApiSource.jikeFree:
        return loc.settings_jike_title;
    }
  }

  String localizedDescription(AppLocalizations loc) {
    switch (id) {
      case ApiSource.googleBooks:
        return loc.settings_google_subtitle;
      case ApiSource.openLibrary:
        return loc.settings_openlibrary_subtitle;
      case ApiSource.wikipedia:
        return loc.settings_wikipedia_subtitle;
      case ApiSource.jikeFree:
        return loc.settings_jike_subtitle;
    }
  }
}

class ApiSourceRegistry {
  static const List<ApiSourceInfo> all = [
    ApiSourceInfo(
      id: ApiSource.googleBooks,
      displayName: 'Google Books',
      description: '',
      enabledByDefault: true,
      requiresKey: false,
      baseUrl: 'https://www.googleapis.com/books/v1/volumes?q=isbn=',
    ),
    ApiSourceInfo(
      id: ApiSource.openLibrary,
      displayName: 'Open Library',
      description: '',
      enabledByDefault: true,
      requiresKey: false,
      baseUrl: 'https://openlibrary.org/api/books?bibkeys=ISBN:',
    ),
    ApiSourceInfo(
      id: ApiSource.wikipedia,
      displayName: 'Wikipedia',
      description: '',
      enabledByDefault: false,
      requiresKey: false,
      baseUrl:
          'https://en.wikipedia.org/w/api.php?action=query&list=search&srsearch=',
    ),
    ApiSourceInfo(
      id: ApiSource.jikeFree,
      displayName: 'Jike Free',
      description: '',
      enabledByDefault: false,
      requiresKey: false,
      baseUrl: 'https://api.jike.xyz/situ/book/isbn/',
    )
  ];

  static ApiSourceInfo info(ApiSource source) {
    return all.firstWhere((item) => item.id == source);
  }

  static String localizedName(AppLocalizations loc, ApiSource source) {
    switch (source) {
      case ApiSource.googleBooks:
        return loc.settings_google_title;
      case ApiSource.openLibrary:
        return loc.settings_openlibrary_title;
      case ApiSource.wikipedia:
        return loc.settings_wikipedia_title;
      case ApiSource.jikeFree:
        return loc.settings_jike_title;
    }
  }

  static String localizedDescription(AppLocalizations loc, ApiSource source) {
    switch (source) {
      case ApiSource.googleBooks:
        return loc.settings_google_subtitle;
      case ApiSource.openLibrary:
        return loc.settings_openlibrary_subtitle;
      case ApiSource.wikipedia:
        return loc.settings_wikipedia_subtitle;
      case ApiSource.jikeFree:
        return loc.settings_jike_subtitle;
    }
  }

  static List<ApiSource> defaultEnabled() {
    return all
        .where((item) => item.enabledByDefault)
        .map((item) => item.id)
        .toList();
  }

  static ApiSource? fromId(String raw) {
    try {
      return ApiSource.values.firstWhere((s) => s.name == raw);
    } catch (_) {
      return null;
    }
  }

  static String toId(ApiSource source) => source.name;
}

File: ios/ISBNManager/Models/ApiSource.swift
import Foundation

/// ISBN 查詢來源。
enum ApiSource: String, CaseIterable, Identifiable {
    case googleBooks = "googleBooks"
    case openLibrary = "openLibrary"
    case wikipedia = "wikipedia"
    case jikeFree = "jikeFree"
    case wikidata = "wikidata"
    case libraryOfCongress = "libraryOfCongress"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .googleBooks: return "Google Books"
        case .openLibrary: return "Open Library"
        case .wikipedia: return "Wikipedia"
        case .jikeFree: return "Jike Free"
        case .wikidata: return "Wikidata"
        case .libraryOfCongress: return "Library of Congress"
        }
    }

    var enabledByDefault: Bool {
        switch self {
        case .googleBooks, .openLibrary, .wikidata, .libraryOfCongress: return true
        case .wikipedia, .jikeFree: return false
        }
    }

    var requiresKey: Bool {
        switch self {
        case .googleBooks, .jikeFree: return true
        case .openLibrary, .wikipedia, .wikidata, .libraryOfCongress: return false
        }
    }

    /// 使用者填入 API key 的儲存鍵（若需要 key）。nil 表示不需 key。
    var keyStorageKey: String? {
        switch self {
        case .googleBooks: return "google_books_api_key"
        case .jikeFree: return "jike_api_key"
        default: return nil
        }
    }

    var baseUrl: String {
        switch self {
        case .googleBooks: return "https://www.googleapis.com/books/v1/volumes?q=isbn="
        case .openLibrary: return "https://openlibrary.org/api/books?bibkeys=ISBN:"
        case .wikipedia: return "https://en.wikipedia.org/w/api.php?action=query&list=search&srsearch="
        case .jikeFree: return "https://api.jike.xyz/situ/book/isbn/"
        case .wikidata: return "https://query.wikidata.org/sparql"
        case .libraryOfCongress: return "https://www.loc.gov/search/"
        }
    }

    /// 本地化名稱。
    func localizedName(_ s: Strings) -> String {
        switch self {
        case .googleBooks: return s.t("settings_google_title")
        case .openLibrary: return s.t("settings_openlibrary_title")
        case .wikipedia: return s.t("settings_wikipedia_title")
        case .jikeFree: return s.t("settings_jike_title")
        case .wikidata: return s.t("settings_wikidata_title")
        case .libraryOfCongress: return s.t("settings_loc_title")
        }
    }

    /// 本地化描述。
    func localizedDescription(_ s: Strings) -> String {
        switch self {
        case .googleBooks: return s.t("settings_google_subtitle")
        case .openLibrary: return s.t("settings_openlibrary_subtitle")
        case .wikipedia: return s.t("settings_wikipedia_subtitle")
        case .jikeFree: return s.t("settings_jike_subtitle")
        case .wikidata: return s.t("settings_wikidata_subtitle")
        case .libraryOfCongress: return s.t("settings_loc_subtitle")
        }
    }

    static var all: [ApiSource] { ApiSource.allCases }
    static func defaultEnabled() -> [ApiSource] { allCases.filter { $0.enabledByDefault } }
    static func fromId(_ raw: String) -> ApiSource? {
        ApiSource(rawValue: raw)
    }
}
